File: Javascript/Arrays/Problem17.js
 

 // 1. Brute Force solution

 for(let i=0;i<arr.length;i++){
    let count=0;
    let temp=arr[i];
    for(let j=i;j<arr.length;j++){
      if(temp==arr[j]){
        count++;
      }
    }
    if(count>=arr.length/2){
      console.log(temp);
      break;
    }
  }

// 2 . Better solution 

let arr=[4,4,2,4,3,4,4,3,2,4];

let map=new Map(); // map(key,val)


for(let i=0;i<arr.length;i++){
  if(map.has(arr[i])){
    let val=map.get(arr[i]);
    map.set(arr[i],val+1);
  }
  else{
    map.set(arr[i],1);
  }
}


for(let [key,val] of map){
  if(val>=arr.length/2){
    console.log(key);
    break;
  }
}




// 3 Optimal Solution 









let ctn=0

let el;

for(let i=0;i<arr.length;i++){
  if(ctn==0){
    ctn=1;
    el=arr[i];
  }
  else if(arr[i]==el){
    ctn++;
  }
  else{
    ctn--;
  }
}

let ctn1=0;

for(let i=0;i<arr.length;i++){
  if(arr[i]==el){
  ctn1++;
  }
}



if(ctn1>=(arr.length/2)){
  console.log(el)
}

File: C++/Arrays/Problem17.cpp


 // 1 . Brute Force solution 



 #include <iostream>
using namespace std;

int main() 
{
    int n=6;
    int arr[n]={3,2,3};
    
    for(int i=0;i<n;i++){
      int temp=arr[i];
      int ctn=0;
      for(int j=i;j<n;j++){
        if(temp==arr[i]){
          ctn++;
        }
      }
      if(ctn>=n/2){
        std::cout << temp << std::endl;
        break;
      }
    }
    
    
    return 0;
}

// 2 Better solution


#include <iostream>
#include <bits/stdc++.h>
using namespace std;

int main() 
{
    int n=6;
    int arr[n]={3,2,3};
    
    map<int,int>hashmap;
    
    for(int i=0;i<n;i++){
     hashmap[arr[i]]++;
    }
    
    for(auto it :hashmap){
     if(it.first>=n/2){
       std::cout << it.second << std::endl;
     }
    }
    return 0;
}


 // 3 Better solution 


 #include <iostream>
#include <bits/stdc++.h>
using namespace std;

int main() 
{
    int n=3;
    int arr[n]={3,2,3};
    int ctn=0;
    int el;
    for(int i=0;i<n;i++){
      if(ctn==0){
        ctn=1;
        el=arr[i];
      }
      else if(el==arr[i]){
        ctn++;
      }
      else{
        ctn--;
      }
    }
    
    int ctn1=0;
    
    for(int i=0;i<n;i++){
      if(el==arr[i]){
        ctn1++;
      }
    }
    
    if(ctn1>=n/2){
      std::cout << el << std::endl;
    }
    return 0;
}